### DISCUSSION TITLE
Key-Value Separation in LSM Storage Engines

### DISCUSSION THREAD

**Discussion by skyzh** [skyzh/skyzh-site#28] [General]
https://skyzh.dev/blog/2023-12-31-lsm-kv-separation-overview/

**Comment by artyomVB** [comment]
Do I understand correctly that exceeding garbage ratio or age_cutoff triggers compaction in BlobDB?  Or does compaction run independently and only select victim files based on these parameters?

**Comment by manishrjain** [comment]
Just read this blog post (for the 1st time). The way Badger figures out about the discard stats is not this:

> As shown in the above figure, BadgerDB stores the garbage estimation value of each vLog in the DISCARD file. When a user updates or deletes a key, the counter of the corresponding vLog file will be incremented by 1. When the counter reaches a certain threshold, the vLog file will be rewritten.

Instead, it tracks which keys are being discarded during compactions. It looks at the value of these keys, and if it's a value log pointer, it can update the discard stats for that file.

https://github.com/hypermodeinc/badger/blob/main/levels.go#L651-L664

The other thing I didn't understand why you mentioned that doing a range scan is "slow" with Badger, but not with others. Why's that the case? value prefetching is an optimization, not a flaw. It's not required for Badger to work. And every LSM tree with key value separation needs to fetch the key, then look at the value pointer, then fetch the value -- same work across the board.

**Reply by skyzh** [reply]
The blog post was written 5 years ago, so either I made a mistake or things have changed in the past 5 years; I will fix it.

Is Badger vLog sorted? I assume no. Other key-value separated LSM implementations sorts the vLog and therefore they are more likely to leverage the spatial locality of the data and do fewer I/O during the scan. I did some benchmark back to the time I wrote the blog post and from what I can remember, range scan takes longer on Badger.

**Reply by manishrjain** [reply]
> The blog post was written 5 years ago

The date shows Dec 31, 2023.

The vlog doesn't need to be sorted. The pointer stored in the LSM tree has the file id, the offset and the length of the value. So, it's a simple fetch from mmap / file.

No DB should be doing binary searches over the value log anyway, because it doesn't contain all the keys. It should only contain the key-values where the value exceeds a certain threshold -- 4KB as you mentioned in the blog.

**Reply by skyzh** [reply]
> it's a simple fetch from mmap / file.

Yes, it's supposed to read (file_id, offset, length), and I'm saying that if the vlog is sorted, range scan gets better locality, because the engine will likely fetch a continuous range of bytes within some vlog SST file. I'm not suggesting that other DBs are doing binary searches.

> The date shows Dec 31, 2023.

It was originally posted in 2021 in Chinese and I did a ChatGPT translation in 2023. So it might not reflect latest status of these projects. But I'll read the code again and update the blog.
